# NLP Analysis

In this notebook, we are going to apply NLP techniques to discover what makes a Linkedin post good.

We have created several datasets to make this analysis
* contentCorpus.pkl
* sinekDtm.pkl

## Libraries

In [2]:
import pandas as pd

## Content analysis

In this section, let's analyze the content of each author.  
We can perform some analysis :
* which words are correlated with a high attractivity ?
* sentiment analysis
* topic modeling

### Word's attractivity

Let's see if there are specific words that are associated with the number of reactions and comments.  
We pick Simek Sinek's content only.  
Therefore it allows us to analyse content without bias. 

In [20]:
#We need a Document Term Matrix
#Let's take the one we created about Simon Sinek
df = pd.read_pickle("sinekDtm.pkl")
df.head(2)

#Reactions  #Comments                                            Content  \
0       12093        257  we are only in charge when we are willing to l...   
1        5415        164  when the people have to manage dangers from in...   

   ability  abitofoptimism  able  abnormal  absolutely  accept  accepting  \
0        0               0     0         0           0       0          0   
1        0               0     1         0           0       0          0   

   ...  years  yesterday  yin  york  youll  youre  youtube  \
0  ...      0          0    0     0      0      0        0   
1  ...      0          0    0     0      0      0        0   

   youtubecomsimonsinek  youve  única  
0                     0      0      0  
1                     0      0      0  

[2 rows x 1533 columns]

In [78]:
df.shape

(245, 1533)

In [27]:
df.dtypes

#Reactions               int64
#Comments                int64
Content                 object
ability                  int64
abitofoptimism           int64
                         ...  
youre                    int64
youtube                  int64
youtubecomsimonsinek     int64
youve                    int64
única                    int64
Length: 1533, dtype: object

We note 245 posts from Simon Sinek, and 1530 different words.

Now, to observe which words have the most influence on Reactions & Comments, we are going to process as the following : 
* If a word is present in a post, it took as value the number of Reactions & Comments associated with this post
* We count in how many posts each word is present (if a word appears several times in a post, it's counted as 1)
* For each word, we calculate the mean
* We classify these words

In [ ]:
#Transform the Dtm to associate a word presence with the number of reactions (only)

#We create the matrix of number of reactions for each post
reactionsMatrix = pd.DataFrame(columns=df.columns[3:]) #Columns == words from document-term matrix


for i in range(3,df.shape[1]): #We iterate on each word column
    reactionsMatrix.iloc[:,i-3] = df.iloc[:,i].apply(lambda x : df.iloc[x,0] if x>0 else 0) #if x>0, it takes the number of reactions

In [90]:
for i in range(3,df.shape[1]): #We iterate on each word column
    reactionsMatrix.iloc[:,i-3] = df.apply(lambda x : print(x))

0      12093
1       5415
2        706
3       5934
4      45732
       ...  
240     4005
241     1698
242      661
243      766
244      789
Name: #Reactions, Length: 245, dtype: int64
0       257
1       164
2        24
3       310
4      1200
       ... 
240      93
241      74
242      59
243      35
244      23
Name: #Comments, Length: 245, dtype: int64
0      we are only in charge when we are willing to l...
1      when the people have to manage dangers from in...
2      where do our opinions come from could our assu...
3      if you have an infinite mindset but your leade...
4      if you want to feel happy do something for you...
                             ...                        
240    igniter of the year  well i know that im an op...
241    executives who prioritize the shareholder are ...
242    like many i too have been reflecting as we nea...
243    if you say customer first that means your empl...
244    the small work hard to serve themselves in a b...
Name: Conte

IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed

In [83]:
reactionsMatrix

0      12093
1       5415
2        706
3       5934
4      45732
       ...  
240     4005
241     1698
242      661
243      766
244      789
Length: 245, dtype: int64